File: src/main/java/shopping/dto/request/ProductRequest.java
package shopping.dto.request;

import static shopping.dto.request.validator.RequestArgumentValidator.validateNumberNotNullAndPositive;
import static shopping.dto.request.validator.RequestArgumentValidator.validateStringNotNullAndNotBlankAndNotLong;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;

public class ProductRequest {

    private static final int MAX_NAME_LENGTH = 255;
    private static final String NAME_NAME = "name";
    private static final String IMAGE_URL_NAME = "imageUrl";
    private static final String PRICE_NAME = "price";

    private final String name;
    private final String imageUrl;
    private final Long price;

    @JsonCreator
    public ProductRequest(@JsonProperty(NAME_NAME) final String name,
            @JsonProperty(IMAGE_URL_NAME) final String imageUrl,
            @JsonProperty(PRICE_NAME) final Long price) {
        validate(name, imageUrl, price);
        this.name = name;
        this.imageUrl = imageUrl;
        this.price = price;
    }

    private void validate(String name, String imageUrl, Long price) {
        validateStringNotNullAndNotBlankAndNotLong(name, NAME_NAME, MAX_NAME_LENGTH);
        validateStringNotNullAndNotBlankAndNotLong(imageUrl, IMAGE_URL_NAME, MAX_NAME_LENGTH);
        validateNumberNotNullAndPositive(price, PRICE_NAME);
    }

    public String getName() {
        return name;
    }

    public String getImageUrl() {
        return imageUrl;
    }

    public Long getPrice() {
        return price;
    }
}

File: src/main/kotlin/shopping/controller/dto/ProductRequest.kt
package shopping.controller.dto

import com.fasterxml.jackson.annotation.JsonProperty
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotNull
import jakarta.validation.constraints.Positive
import shopping.domain.Product

data class ProductRequest(
    @field:NotBlank
    @JsonProperty("name")
    private val _name: String?,
    @field:Positive
    @field:NotNull
    @JsonProperty("price")
    private val _price: Long?,
    @field:NotBlank
    @JsonProperty("imageUrl")
    private val _imageUrl: String?,
) {
    val name: String get() = _name!!
    val price: Long get() = _price!!
    val imageUrl: String get() = _imageUrl!!

    fun toDomain(): Product = Product(name, price, imageUrl)
}
